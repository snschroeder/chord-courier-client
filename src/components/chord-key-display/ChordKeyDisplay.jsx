import React from 'react';
import PropTypes from 'prop-types';
import ChordGallery from '../chord-display/chord-gallery/ChordGallery';

export default function ChordKeyDisplay(props) {
  const { chords } = props;

  const indices = [];
  const progSortedChords = [];
  if (chords.length !== 0) {
    chords.sort((a, b) => {
      if (a.notation[0] > b.notation[0]) return 1;
      if (a.notation[0] < b.notation[0]) return -1;
      if (a.variation > b.variation) return 1;
      return -1;
    });
    const splitIndex = chords.findIndex((chord) => chord.notation === chord.tonic);
    const prefix = chords.splice(0, splitIndex);
    chords.push(...prefix);

    for (let i = 0; i < chords.length - 2; i += 1) {
      if (chords[i].notation !== chords[i + 1].notation) {
        indices.push(i + 1);
      }
    }
    indices.unshift(0);
    indices.push(chords.length - 1);

    for (let i = 0; i < indices.length; i += 1) {
      progSortedChords.push(chords.slice(indices[i], indices[i + 1]));
    }
  }

  return (
    <ul className="grid">
      {progSortedChords.map((chordGroup, index) => (
        // eslint-disable-next-line react/no-array-index-key
        <li key={index} className="grid-item">
          <ChordGallery chords={chordGroup} />
        </li>

      ))}
    </ul>
  );
}

ChordKeyDisplay.defaultProps = {
  chords: [],
};

ChordKeyDisplay.propTypes = {
  chords: PropTypes.instanceOf(Array),
};
