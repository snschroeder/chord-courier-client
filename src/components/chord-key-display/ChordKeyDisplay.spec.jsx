/* eslint-disable no-undef */
import React from 'react';
import ReactDOM from 'react-dom';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { BrowserRouter } from 'react-router-dom';
import ChordKeyDisplay from './ChordKeyDisplay';
import testArrays from '../../testHelpers';

describe('ChordDisplay component', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(
      <BrowserRouter>
        <ChordKeyDisplay chords={testArrays.fullCMaj} />
      </BrowserRouter>,
      div,
    );
    ReactDOM.unmountComponentAtNode(div);
  });
  it('renders the display correctly', () => {
    const wrapper = shallow(<ChordKeyDisplay chords={testArrays.fullCMaj} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
