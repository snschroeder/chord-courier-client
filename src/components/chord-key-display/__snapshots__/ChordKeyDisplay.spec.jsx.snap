// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChordDisplay component renders the display correctly 1`] = `
<ul
  className="grid"
>
  <li
    className="grid-item"
    key="0"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,x,2,1,o,0,o",
            "chord_name": "C major",
            "fifth": "G",
            "fourth": "F",
            "id": 3,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, beginner, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "0,x,2,1,o,2,2",
            "chord_name": "C major add9",
            "fifth": "G",
            "fourth": "F",
            "id": 39,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "jazz, open",
            "third": "Em",
            "tonic": "C",
            "variation": "2",
          },
          Object {
            "chord_code": "3,x,0,2,o,o,o",
            "chord_name": "C major 7",
            "fifth": "G",
            "fourth": "F",
            "id": 40,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "dissonant, open, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
          Object {
            "chord_code": "8,0,2,1,1,0,0,b",
            "chord_name": "C major 7",
            "fifth": "G",
            "fourth": "F",
            "id": 44,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "4",
          },
          Object {
            "chord_code": "3,0,2,2,x,x,x",
            "chord_name": "C5",
            "fifth": "G",
            "fourth": "F",
            "id": 59,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, rock, blues, power chord",
            "third": "Em",
            "tonic": "C",
            "variation": "5",
          },
          Object {
            "chord_code": "0,o,2,x,o,2,o",
            "chord_name": "Cadd9",
            "fifth": "G",
            "fourth": "F",
            "id": 65,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, rock, blues",
            "third": "Em",
            "tonic": "C",
            "variation": "6",
          },
          Object {
            "chord_code": "0,2,2,1,o,o,o",
            "chord_name": "CMaj7/G",
            "fifth": "G",
            "fourth": "F",
            "id": 74,
            "notation": "C",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "7",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="1"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,x,x,o,1,2,0",
            "chord_name": "D minor",
            "fifth": "G",
            "fourth": "F",
            "id": 10,
            "notation": "Dm",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, beginner, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "0,x,o,2,1,2,o",
            "chord_name": "Dmin add9/A",
            "fifth": "G",
            "fourth": "F",
            "id": 82,
            "notation": "Dm",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "2",
          },
          Object {
            "chord_code": "5,x,o,2,0,1,o",
            "chord_name": "Dmin9/A",
            "fifth": "G",
            "fourth": "F",
            "id": 84,
            "notation": "Dm",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
          Object {
            "chord_code": "5,x,0,2,0,1,2,b",
            "chord_name": "Dmin13",
            "fifth": "G",
            "fourth": "F",
            "id": 91,
            "notation": "Dm",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, jazz, pop, indie",
            "third": "Em",
            "tonic": "C",
            "variation": "4",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="2"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,o,1,1,o,o,o",
            "chord_name": "E minor",
            "fifth": "G",
            "fourth": "F",
            "id": 11,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, beginner, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "0,o,1,1,o,2,2",
            "chord_name": "E minor 7",
            "fifth": "G",
            "fourth": "F",
            "id": 38,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "jazz, open, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "2",
          },
          Object {
            "chord_code": "0,o,1,o,o,2,1",
            "chord_name": "E minor 9",
            "fifth": "G",
            "fourth": "F",
            "id": 41,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "dissonant, open",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
          Object {
            "chord_code": "0,o,x,1,3,2,2",
            "chord_name": "E minor 7",
            "fifth": "G",
            "fourth": "F",
            "id": 45,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, seventh, dark, brooding",
            "third": "Em",
            "tonic": "C",
            "variation": "4",
          },
          Object {
            "chord_code": "5,x,2,0,o,o,o",
            "chord_name": "E minor",
            "fifth": "G",
            "fourth": "F",
            "id": 78,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "5",
          },
          Object {
            "chord_code": "5,o,2,0,2,3,0,b",
            "chord_name": "Emin11",
            "fifth": "G",
            "fourth": "F",
            "id": 71,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced",
            "third": "Em",
            "tonic": "C",
            "variation": "5",
          },
          Object {
            "chord_code": "12,o,0,0,0,o,o",
            "chord_name": "Emin11",
            "fifth": "G",
            "fourth": "F",
            "id": 72,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced",
            "third": "Em",
            "tonic": "C",
            "variation": "6",
          },
          Object {
            "chord_code": "7,x,o,2,0,1,o",
            "chord_name": "Emin7/A",
            "fifth": "G",
            "fourth": "F",
            "id": 85,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "6",
          },
          Object {
            "chord_code": "7,x,0,2,0,1,0,b",
            "chord_name": "Emin7",
            "fifth": "G",
            "fourth": "F",
            "id": 113,
            "notation": "Em",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced, jazz, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "7",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="3"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,0,2,2,1,0,0,b",
            "chord_name": "F major",
            "fifth": "G",
            "fourth": "F",
            "id": 6,
            "notation": "F",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "7,x,1,0,2,1,x",
            "chord_name": "FMaj9",
            "fifth": "G",
            "fourth": "F",
            "id": 88,
            "notation": "F",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, jazz, pop, indie",
            "third": "Em",
            "tonic": "C",
            "variation": "2",
          },
          Object {
            "chord_code": "0,x,o,2,1,0,0,b",
            "chord_name": "FMaj/A",
            "fifth": "G",
            "fourth": "F",
            "id": 120,
            "notation": "F",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, barre",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="4"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,2,1,o,o,2,2",
            "chord_name": "G major",
            "fifth": "G",
            "fourth": "F",
            "id": 7,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, beginner, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "10,x,x,2,o,2,o",
            "chord_name": "G major",
            "fifth": "G",
            "fourth": "F",
            "id": 43,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
          Object {
            "chord_code": "0,2,1,o,o,o,0",
            "chord_name": "G major 7",
            "fifth": "G",
            "fourth": "F",
            "id": 46,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "4",
          },
          Object {
            "chord_code": "7,x,x,o,0,1,0",
            "chord_name": "G major/D",
            "fifth": "G",
            "fourth": "F",
            "id": 52,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "5",
          },
          Object {
            "chord_code": "3,0,2,2,x,x,x",
            "chord_name": "G5",
            "fifth": "G",
            "fourth": "F",
            "id": 58,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, rock, blues, power chord",
            "third": "Em",
            "tonic": "C",
            "variation": "6",
          },
          Object {
            "chord_code": "3,0,2,1,3,o,x",
            "chord_name": "GMaj7 add#11",
            "fifth": "G",
            "fourth": "F",
            "id": 69,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "7",
          },
          Object {
            "chord_code": "0,2,x,o,1,2,2",
            "chord_name": "G5 add9",
            "fifth": "G",
            "fourth": "F",
            "id": 75,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, power chord",
            "third": "Em",
            "tonic": "C",
            "variation": "8",
          },
          Object {
            "chord_code": "7,0,3,2,0,0,0,b",
            "chord_name": "GMaj7/B",
            "fifth": "G",
            "fourth": "F",
            "id": 92,
            "notation": "G",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "9",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="5"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,x,o,1,1,0,o",
            "chord_name": "A minor",
            "fifth": "G",
            "fourth": "F",
            "id": 8,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "basic, beginner, open",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
          Object {
            "chord_code": "0,x,o,1,o,2,2",
            "chord_name": "A minor 7 sus4",
            "fifth": "G",
            "fourth": "F",
            "id": 37,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "jazz, open, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "2",
          },
          Object {
            "chord_code": "5,x,o,0,0,0,2,b",
            "chord_name": "A minor 9",
            "fifth": "G",
            "fourth": "F",
            "id": 53,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate, jazz, mellow",
            "third": "Em",
            "tonic": "C",
            "variation": "3",
          },
          Object {
            "chord_code": "4,x,o,0,1,1,3",
            "chord_name": "Amin#6 add9",
            "fifth": "G",
            "fourth": "F",
            "id": 70,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced",
            "third": "Em",
            "tonic": "C",
            "variation": "4",
          },
          Object {
            "chord_code": "12,0,0,2,o,1,3",
            "chord_name": "Amin7/E",
            "fifth": "G",
            "fourth": "F",
            "id": 77,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced, seventh",
            "third": "Em",
            "tonic": "C",
            "variation": "5",
          },
          Object {
            "chord_code": "5,x,o,0,1,1,o",
            "chord_name": "Amin6",
            "fifth": "G",
            "fourth": "F",
            "id": 83,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "6",
          },
          Object {
            "chord_code": "9,x,o,o,1,2,0",
            "chord_name": "A minor",
            "fifth": "G",
            "fourth": "F",
            "id": 86,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "intermediate",
            "third": "Em",
            "tonic": "C",
            "variation": "7",
          },
          Object {
            "chord_code": "5,x,o,1,0,0,o",
            "chord_name": "Amin/Maj7",
            "fifth": "G",
            "fourth": "F",
            "id": 93,
            "notation": "Am",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "advanced, seventh, jazz",
            "third": "Em",
            "tonic": "C",
            "variation": "8",
          },
        ]
      }
    />
  </li>
  <li
    className="grid-item"
    key="6"
  >
    <ChordGallery
      chords={
        Array [
          Object {
            "chord_code": "0,x,1,2,3,2,x",
            "chord_name": "B diminished",
            "fifth": "G",
            "fourth": "F",
            "id": 13,
            "notation": "Bdim",
            "second": "Dm",
            "seventh": "Bdim",
            "sixth": "Am",
            "tag": "diminished",
            "third": "Em",
            "tonic": "C",
            "variation": "1",
          },
        ]
      }
    />
  </li>
</ul>
`;
